src/util/numerics.js

```diff
@@ -0,0 +1,70 @@
+// @flow
+
+import * as C from "./combo";
+
+/**
+ * A library for making type-level assertions about numeric properties;
+ * e.g. that a number is finite, or finite and non-negative, or an integer.
+ *
+ * We should avoid having these methods in numeric hot loops by doing "unsafe"
+ * operations in the hot loop, and then validating on exiting the loop.
+ *
+ * Nonetheless, we avoid unnecessary function calls for performance reasons (i.e.
+ * the validator functions avoid calling each other).
+ */
+
+// A number guaranteed to not be NaN, Infinity, or -Infinity
+export opaque type Finite: number = number;
+export function finite(n: number): Finite {
+  if (!isFinite(n)) {
+    throw new Error(`expected finite, got: ${n}`);
+  }
+  return n;
+}
+export const finiteParser: C.Parser<Finite> = C.fmap(C.number, finite);
+
+export opaque type FiniteNonnegative: number = number;
+export function finiteNonnegative(n: number): FiniteNonnegative {
+  if (n < 0 || !isFinite(n)) {
+    throw new Error(`expected finite nonnegative, got: ${n}`);
+  }
+  return n;
+}
+export const finiteNonnegativeParser: C.Parser<FiniteNonnegative> = C.fmap(
+  C.number,
+  finiteNonnegative
+);
+
+export opaque type Integer: number = number;
+export function integer(n: number): Integer {
+  if (!isFinite(n) || n !== Math.floor(n)) {
+    throw new Error(`expected integer, got: ${n}`);
+  }
+  return n;
+}
+export const integerParser: C.Parser<Integer> = C.fmap(C.number, integer);
+
+export opaque type NonnegativeInteger: number = number;
+export function nonnegativeInteger(n: number): NonnegativeInteger {
+  if (!isFinite(n) || n !== Math.floor(n) || n < 0) {
+    throw new Error(`expected nonnegative integer, got: ${n}`);
+  }
+  return n;
+}
+export const nonnegativeIntegerParser: C.Parser<NonnegativeInteger> = C.fmap(
+  C.number,
+  nonnegativeInteger
+);
+
+// A number guaranteed to be in the range [0, 1]
+export opaque type Proportion: number = number;
```

Comment:
So this is an interesting one. :-)

First, I don’t think that _proportion_ is the right name. From context
in this module, I’d expect the name _proportion_ to be used in the
mathematical sense, but I can’t link that to its usage here. Constants
of proportionality can be greater than one, or negative. _Ratios_ can
also of course be greater than one.

Now, from context broader than this PR, I expect that you intend to use
this _proportion_ more like _share_ (“…of the pie”). In that case, it
seems that actually _portion_ (not _proportion_) fits quite nicely? (It
actually just seems more fitting to me overall; non-mathematically,
proportions are about how things relate to each other, like aspect
ratios or measurements, whereas portions are the things that get divvied
up.)

That said, I wonder whether you’ll end up wanting this type at all.
For validating portions, won’t you always want to verify that the
distribution sums to unity, or at least does not exceed unity? in which
case it suffices to take `FiniteNonnegative`s and assert on the sum.

Reply:
I like "portion", but if we want a more "official term" for this, could use "proper fraction" ? Still prefer "portion" though